app/api/events.py

```diff
@@ -586,7 +619,229 @@ async def confirmation(request: Request, id: str, token: AccessTokenPayload = De
+    if not user_event or 'participants' not in user_event:
+        raise HTTPException(400, "User not joined event")
+    
+    # Update attendence
+    res = db.events.update_one(
+        {'eid': event['eid'], 'participants': {
+            '$elemMatch': {'id': member['id']}
+        }},
+        {'$set': {'participants.$.attended': True}}
+    )
+
+    if not res:
+        raise HTTPException(500)
+
+
+@router.put('/{id}/register', dependencies=[Depends(validate_uuid)])
+def set_attendence(request: Request, id: str, payload: SetAttendencePayload, token: AccessTokenPayload = Depends(authorize_admin)):
+    """ Set attendence of memeber for event """
+    db = get_database(request)
+    event = get_event_or_404(db, id)
+    member = db.members.find_one({'id': UUID(payload.member)})
+
+    # Check whether member id in payload is valid
+    if not member:
+        raise HTTPException(404, "User could not be found")
+    
+    # Get event data (if member is joined event)
+    user_event = db.events.find_one({"eid": event["eid"]}, {"participants": {
+        "$elemMatch": {"id": member["id"]}}})
+    
+    # Check whether member exists in event
+    if not user_event or 'participants' not in user_event:
+        raise HTTPException(400, "User not joined event")
+    
+    # Update attendence
+    res = db.events.update_one(
+        {'eid': event['eid'], 'participants': {
+            '$elemMatch': {'id': member['id']}
+        }},
+        {'$set': {'participants.$.attended': payload.attendence}}
+    )
+
+    if not res:
+        raise HTTPException(500)
+    
+
+@router.post('/{id}/registerAbsence', dependencies=[Depends(validate_uuid)])
+async def register_absence(request: Request, id: str, token: AccessTokenPayload = Depends(authorize_admin)):
+    """ Give all members absent on binding event a penalty """
+    db = get_database(request)
+    event = get_event_or_404(db, id)
+
+    if not event["bindingRegistration"]:
+        raise HTTPException(400, "Cannot penalize on non-binding events")
+    
+    if not event_has_started(event):
+        raise HTTPException(400, "Cannot penalize before event start") 
+    
+    if not event["confirmed"]:
+        raise HTTPException(400, "Cannot penalize unconfirmed event")
+        
+
+    # Aggregate all participants that are confirmed, but not attended,
+    # and group them by ids
+    pipeline = [
+        {"$match": {"eid": event["eid"]}},
+        {"$unwind": {"path": "$participants"}},
+        {"$match": {"participants.confirmed": {"$eq": True}}},
+        {"$match": {"participants.attended": {"$ne": True}}},
+        {"$group": {"_id": "$participants.id"}},
+    ]
+    absent = list(db.events.aggregate(pipeline))
+
+    if not absent or len(absent) == 0:
+        raise HTTPException(400, "No members to penalize")
+    
+    # Get ids
+    absent_ids = [p["_id"] for p in absent]
+    
+    # Get already penalized ids
+    penalized_ids = event["registeredPenalties"]
```

Comment:
Correct, this variable is used to remove the already penalized members from the queue

Reply:
I see, an idea could be: 
```python3
for m_id in absent_ids:
        if m_id not in penalized_ids:
            absent_ids.remove(m_id)
```
to avoid creating an extra list

Reply:
That would be more elegant.
Does this definitely work? I remember having troubles removing items from a list I was iterating through in python another time. Will investigate.

Reply:
if `in` triggers `remove` should get a match